File: Sources/XMTP/Messages/PagingInfo.swift
//
//  PagingInfo.swift
//
//
//  Created by Pat Nakajima on 12/15/22.
//

import Foundation

typealias PagingInfo = Xmtp_MessageApi_V1_PagingInfo
typealias PagingInfoCursor = Xmtp_MessageApi_V1_Cursor
public typealias PagingInfoSortDirection = Xmtp_MessageApi_V1_SortDirection

public struct Pagination {
	public var limit: Int?
    public var before: Date?
    public var after: Date?
    public var direction: PagingInfoSortDirection?
        
    public init(limit: Int? = nil, before: Date? = nil, after: Date? = nil, direction: PagingInfoSortDirection? = .descending) {
        self.limit = limit
        self.before = before
        self.after = after
        self.direction = direction
    }

	var pagingInfo: PagingInfo {
		var info = PagingInfo()

		if let limit {
			info.limit = UInt32(limit)
		}
        info.direction = direction ?? Xmtp_MessageApi_V1_SortDirection.descending
		return info
	}
}

extension PagingInfo {
	init(limit: Int? = nil, cursor: PagingInfoCursor? = nil, direction: PagingInfoSortDirection? = nil) {
		self.init()

		if let limit {
			self.limit = UInt32(limit)
		}

		if let cursor {
			self.cursor = cursor
		}

		if let direction {
			self.direction = direction
		}
	}
}

File: library/src/main/java/org/xmtp/android/library/messages/PagingInfo.kt
package org.xmtp.android.library.messages

import org.xmtp.proto.message.api.v1.MessageApiOuterClass.Cursor
import org.xmtp.proto.message.api.v1.MessageApiOuterClass.SortDirection
import java.util.Date

typealias PagingInfo = org.xmtp.proto.message.api.v1.MessageApiOuterClass.PagingInfo
typealias PagingInfoCursor = Cursor
typealias PagingInfoSortDirection = SortDirection

data class Pagination(
    val limit: Int? = null,
    val direction: PagingInfoSortDirection? = SortDirection.SORT_DIRECTION_DESCENDING,
    val before: Date? = null,
    val after: Date? = null,
) {
    val pagingInfo: PagingInfo
        get() {
            return PagingInfo.newBuilder().also { page ->
                limit?.let {
                    page.limit = it
                }
                if (direction != null) {
                    page.direction = direction
                }
            }.build()
        }
}
